tests/certification/bindings/cron/cron_test.go

```diff
@@ -0,0 +1,266 @@
+		Step("stop sidecar", sidecar.Stop(sidecarName)).
+		Step("stop app", app.Stop(appName)).
+		Run()
+}
+
+// For cron component with trigger @every 3s, check if the app is invoked correctly on app restart
+func TestCronBindingWithAppRestart(t *testing.T) {
+	ports, _ := dapr_testing.GetFreePorts(3)
+	grpcPort := ports[0]
+	httpPort := ports[1]
+	appPort := ports[2]
+
+	cronName := "cron3s"
+	appName := "cronapp3s"
+	sidecarName := "cron-sidecar"
+
+	// Test if cron triggers 5 times within 15 seconds
+	expectedTriggerCount := 5
+	timeoutToObeserveTrigger := time.Second * 15
+
+	// allow cron to trigger for some time before stopping the app
+	waitBeforeAppStop := time.Second * 5
+
+	// wait for some time after the app has stopped, before restarting the app
+	waitBeforeAppRestart := time.Second * 5
+
+	// wait after delete operation to check if cron still triggers
+	waitAfterDelete := time.Second * 10
+
+	triggerWatcher := watcher.NewOrdered()
+	for i := 1; i <= expectedTriggerCount; i++ {
+		triggerWatcher.ExpectInts(i)
+	}
+
+	// total times cron is triggered
+	triggeredCount := 0
+	// store triggered count at the time of cron delete operation
+	var triggeredCountAtDelete int
+
+	flow.New(t, "test cron trigger schedule @every3s with app restart").
+		Step(app.Run(appName, fmt.Sprintf(":%d", appPort), appWithTriggerCounter(t, cronName, triggerWatcher, &triggeredCount))).
+		Step(sidecar.Run(sidecarName,
+			embedded.WithComponentsPath("./components"),
+			embedded.WithDaprGRPCPort(grpcPort),
+			embedded.WithAppProtocol(runtime.HTTPProtocol, appPort),
+			embedded.WithDaprHTTPPort(httpPort),
+			componentRuntimeOptions(),
+		)).
+		Step("allow cron to trigger once", flow.Sleep(waitBeforeAppStop)).
+		Step("stop app", app.Stop(appName)).
+		Step("wait before app restart", flow.Sleep(waitBeforeAppRestart)).
+		Step(app.Run(appName, fmt.Sprintf(":%d", appPort), appWithTriggerCounter(t, cronName, triggerWatcher, &triggeredCount))).
+		Step("assert cron triggered within deadline", assertTriggerCount(triggerWatcher, timeoutToObeserveTrigger)).
+		Step("delete cron trigger", deleteCronTrigger(t, cronName, sidecarName, &triggeredCount, &triggeredCountAtDelete)).
+		Step("allow cron to trigger if delete operation failed", flow.Sleep(waitAfterDelete)).
+		Step("assert cron did not trigger after delete", assertTriggerDelete(t, &triggeredCountAtDelete, &triggeredCount)).
+		Step("stop sidecar", sidecar.Stop(sidecarName)).
+		Step("stop app", app.Stop(appName)).
+		Run()
+}
+
+// For cron component with trigger @every 3s, check if the app is invoked correctly on sidecar restart
+func TestCronBindingWithSidecarRestart(t *testing.T) {
+	ports, _ := dapr_testing.GetFreePorts(3)
+	grpcPort := ports[0]
+	httpPort := ports[1]
+	appPort := ports[2]
+
+	cronName := "cron3s"
+	appName := "cronapp3s"
+	sidecarName := "cron-sidecar"
+
+	// Test if cron triggers 5 times within 15 seconds
+	expectedTriggerCount := 5
+	timeoutToObeserveTrigger := time.Second * 15
```

Comment:
This is going to be very flaky as the wall clock isn't going to give you the accuracy you need. I am speaking by experience because I wrote some tests in Dapr/dapr that relied on the wall clock to measure if a background goroutine was working as intended..... and they were very flaky and had to rewrite them.

Please use a mock clock such as the clock library to mock the clock and advance it during your test execution: https://github.com/benbjohnson/clock

Reply:
Forked cron library to use mock clock and placed the library under `internal/cron`
I am now using mock clock for the tests